File: Cube Control/serialHelper.c
/*
 * unixHelper.c
 *
 * Copyright 2011 Thomas Buck <xythobuz@me.com>
 * Copyright 2011 Max Nuding <max.nuding@gmail.com>
 * Copyright 2011 Felix Bäder <baeder.felix@gmail.com>
 *
 * This file is part of LED-Cube.
 *
 * LED-Cube is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * LED-Cube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with LED-Cube.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <stdlib.h>
#include <stdio.h>

#ifdef winHelper
#include "helper/winSerial.c"
#else
#include "helper/unixSerial.c"
#endif

char *fileData = NULL;

void removeFromBeginning(size_t size, size_t remove);
size_t readFile(char *path);
size_t getFileSize(FILE *fp);
void usage(char *name);
int parseNumber(char *s);
int power(int num, int pow);
void printFileData(size_t size);
void printSerialPorts(char ** ports);

/*
Return values:
0: Success
1: Usage error
2: Serial Port Error
3: Data File Error
*/

int main(int argc, char *argv[]) {
	size_t length, written;

		// printf("Debugging Worker... Ignore me!\n");

		if (argc < 2) {
			usage(argv[0]);
			return 1;
		}

		if (argv[1][0] == 'p') {

			if (argc != 2) {
				usage(argv[0]);
				return 1;
			}
			char** ports = getSerialPorts();
			printSerialPorts(ports);
			free(ports);
			return 0;

		} else if (argv[1][0] == 'w') {

			if (argc != 4) {
				usage(argv[0]);
				return 1;
			}
			if (serialOpen(argv[2]) != 0) {
				printf("Error: Could not open %s\n", argv[1]);
				return 2;
			}
			// write file to com port
			length = readFile(argv[3]);
			if (length == 0) {
				printf("Error while reading %s\n", argv[2]);
				return 3;
			}
	
			written = serialWrite(fileData, length);
			while (written < length) {
				removeFromBeginning(length, written);
				length -= written;
				written = serialWrite(fileData, length);
			}

		} else if (argv[1][0] == 'r') {

			if (argc != 4) {
				usage(argv[0]);
				return 1;
			}
			if (serialOpen(argv[2]) != 0) {
				printf("Error: Could not open %s\n", argv[1]);
				return 2;
			}
			// Read from com port to file
			if (argc < 5) {
				printf("No size given. ");
				usage(argv[0]);
				return 1;
			}
			
			length = parseNumber(argv[3]);
			fileData = (char *)malloc(length * sizeof(char));
			// fill fileData with serial Port data
			written = 0;
			while (written < length) {
				written = serialRead(fileData + written, length - written);
			}

			printFileData(length);

		} else {
			printf("Unrecognized Option: %s\n", argv[1]);
			usage(argv[0]);
			return 1;
		}

		free(fileData);
		fileData = NULL;
		serialClose();
		return 0;
	// }
}

void removeFromBeginning(size_t size, size_t remove) {
	size_t i;
	char *tmp = (char *)malloc((size - remove) * sizeof(char));

	for (i = 0; i < (size - remove); i++) {
		tmp[i] = fileData[i + remove];
	}
	free(fileData);
	fileData = tmp;
}

size_t readFile(char *path) {
	size_t size, i;
	FILE *fp;

	fp = fopen(path, "r");
	if (!fp) {
		return 0;
	}

	size = getFileSize(fp);
	fileData = (char *)malloc(size * sizeof(char));
	for (i = 0; i < size; i++) {
		fileData[i] = fgetc(fp);
	}

	fclose(fp);
	return size;
}

size_t getFileSize(FILE *fp) {
	size_t size = 0;
	int c;

	fseek(fp, 0, 0); // Set file pointer to beginning
	
	do { // Calculate size
		c = fgetc(fp);
		size++;
	} while (c != EOF);
	size--;
	
	fseek(fp, 0, 0);

	return size;
}

void usage(char *name) {
#ifdef winHelper
	printf("Usage:\n%s r|w|p [COM1 [C:\\file\\to\\send.txt] [sizeToRead]]\n", name);
#else
	printf("Usage:\n%s r|w|p [/dev/SerialPort [/file/to/send] [sizeToRead]]\n", name);
#endif
}

int parseNumber(char *s) {
	int i, size = 0, result = 0;
	while (*(s++) != '\0') {
		size++;
	}
	for (i = 0; i < size; i++) {
		result += (s[i] - '0') * power(10, (size - i));
	}
	return result;
}

int power(int num, int pow) {
	int result = 1;
	while (pow > 0) {
		result *= num;
		pow--;
	}
	return result;
}

void printFileData(size_t size) {
	int i;
	for (i = 0; i < size; i++) {
		printf("%x", fileData[i]);
		if (i % 8 == 0) {
			printf("\n");
		}
	}
}

void printSerialPorts(char ** ports) {
	int i = 0;
	while (ports[i] != NULL) {
		printf("%s\n", ports[i]);
		i++;
	}
}

File: CubeControl/SerialHelper.java
/*
 * SerialHelper.java
 *
 * Copyright 2011 Thomas Buck <xythobuz@me.com>
 * Copyright 2011 Max Nuding <max.nuding@gmail.com>
 * Copyright 2011 Felix Bäder <baeder.felix@gmail.com>
 *
 * This file is part of LED-Cube.
 *
 * LED-Cube is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * LED-Cube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with LED-Cube.  If not, see <http://www.gnu.org/licenses/>.
 */
 /**
 * Implement commands of cube. You can only open one serial port.
 * If you want to communicate with another port, close this one first!
 * @author Thomas Buck
 * @author Max Nuding
 * @author Felix Bäder
 * @version 1.0
 */

import java.util.Date;
import java.util.ArrayList;
import java.util.List;

public class SerialHelper {

	private final short OK = 0x42;
	private final short ERROR = 0x23;

	private Frame frame;

	/**
	 * Create new SerialHelper.
	 * @param serialPort Name of serial port to use.
	 * @param frame Frame to show error messages 
	 * @throws Exception Could not open serial port.
	 */
	public SerialHelper(String serialPort, Frame frame) throws Exception {
		if (HelperUtility.openPort(serialPort) == false) {
			printErrorMessage("Could not open serial port \"" + serialPort + "\"");
			throw new Exception("Could not open serial port \"" + serialPort + "\"");
		}
		this.frame = frame;
	}

	/**
	 * Poll to check if the cube is there...
	 * @return TRUE if cube is connected.
	 */
	public boolean probeForCube() {
		short[] data = new short[1];
		data[0] = OK;
		if (!writeData(data)) {
			printErrorMessage("Timeout sending probe for Cube");
			return false;
		}
		data = readData(1);
		if ((data == null) || (data[0] != OK)) {
			printErrorMessage("No response from cube!");
			if (data == null) {
				System.out.println("Probe was null!");
			} else {
				System.out.println("Probe was " + data[0]);
			}
			return false;
		}
		return true;
	}

	/**
	 * Recieve all animations saved in cube.
	 * @return A cubeWorker populated with the new data or null.
	 */
	public cubeWorker getAnimationsFromCube() {
		Animation[] animations;
		int animationCount, frameCount;
		short[] data, tmp = new short[1];

		// Send download command
		tmp[0] = 'g';
		if (!writeData(tmp)) {
			printErrorMessage("Timout sending Command");
			return null;
		}
		data = readData(1);
		if ((data == null) || (data[0] != OK)) {
			printErrorMessage("Response Error");
			if (data == null) {
				System.out.println("Download was null!");
			} else {
				System.out.println("Download was " + data[0]);
			}
			return null;
		}

		// Get animation count
		data = readData(1);
		if (data == null) {
			printErrorMessage("Response Error.");
			System.out.println("Did not get animation count!");
			return null;
		} else {
			animationCount = data[0];
		}
		tmp[0] = OK;
		if (!writeData(tmp)) {
			printErrorMessage("Timout acknowledging animation count!");
			return null;
		}

		animations = new Animation[animationCount];

		// Get animations
		for (int a = 0; a < animationCount; a++) {
			Animation currentAnim = new Animation();

			// Get number of frames
			data = readData(1);
			if (data == null) {
				printErrorMessage("Response Error");
				System.out.println("Did not get frame count!");
				return null;
			} else {
				frameCount = data[0];
			}
			tmp[0] = OK;
			if (!writeData(tmp)) {
				printErrorMessage("Timout sending response Frame Count.");
				return null;
			}

			// Get frames
			for (int f = 0; f < frameCount; f++) {
				AFrame currentFrame = new AFrame();

				// Get frame duration
				data = readData(1);
				if (data == null) {
					printErrorMessage("Response Error");
					System.out.println("Did not get frame duration!");
					return null;
				} else {
					currentFrame.setTime(data[0]);
				}
				tmp[0] = OK;
				if (!writeData(tmp)) {
					printErrorMessage("Timout sending response Frame Duration");
					return null;
				}

				// Get frame data
				data = readData(64);
				if (data == null) {
					printErrorMessage("Response Error");
					System.out.println("Did not get frame data!");
					return null;
				} else {
					currentFrame.setData(data);
				}
				tmp[0] = OK;
				if (!writeData(tmp)) {
					printErrorMessage("Timout sending response Frame Data");
					return null;
				}

				// Add frame to animation
				currentAnim.setFrame(currentFrame, f);
			}

			// Add animation to animations list
			animations[a] = currentAnim;
		}

		return new cubeWorker(animations, frame);
	}

	/**
	 * Send all animations in a cubeWorker to the Cube.
	 * @param worker cubeWorker that containts data.
	 * @return 0 on success. -1 on error.
	 */
	public int sendAnimationsToCube(cubeWorker worker) {
		short[] data, tmp = new short[1];

		// Send upload command
		tmp[0] = 's';
		if (!writeData(tmp)) {
			printErrorMessage("Timout sending Command");
			return -1;
		}
		data = readData(1);
		if ((data == null) || (data[0] != OK)) {
			printErrorMessage("Response Error Command");
			if (data == null) {
				System.out.println("Response Command was null!");
			} else {
				System.out.println("Response Command was " + data[0]);
			}
			return -1;
		}

		// Send animation count
		tmp[0] = (short)worker.size();
		if (!writeData(tmp)) {
			printErrorMessage("Timeout sending numOfAnimations");
			return -1;
		}
		data = readData(1);
		if ((data == null) || (data[0] != OK)) {
			printErrorMessage("Response Error");
			if (data == null) {
				System.out.println("Response Count was null!");
			} else {
				System.out.println("Response Count was " + data[0]);
			}
			return -1;
		}

		// Send animations
		for (int a = 0; a < worker.size(); a++) {
			// Send frame count
			tmp[0] = (short)worker.getAnimation(a).size();
			if (!writeData(tmp)) {
				printErrorMessage("Timeout sending numOfFrames");
				return -1;
			}
			data = readData(1);
			if ((data == null) || (data[0] != OK)) {
				printErrorMessage("Response Error");
				if (data == null) {
				System.out.println("Frame Count was null!");
			} else {
				System.out.println("Frame Count was " + data[0]);
			}
				return -1;
			}

			// Send frames
			for (int f = 0; f < worker.getAnimation(a).size(); f++) {
				// Frame duration
				tmp[0] = worker.getAnimation(a).getFrame(f).getTime();
				if (!writeData(tmp)) {
					printErrorMessage("Timeout sending Frame duration");
					return -1;
				}
				data = readData(1);
				if ((data == null) || (data[0] != OK)) {
					printErrorMessage("Response Error");
					if (data == null) {
						System.out.println("Duration was null!");
					} else {
						System.out.println("Duration was " + data[0]);
					}
					return -1;
				}

				// Frame data
				if (!writeData(worker.getAnimation(a).getFrame(f).getData())) {
					printErrorMessage("Timeout sending Frame");
					return -1;
				}
				data = readData(1);
				if ((data == null) || (data[0] != OK)) {
					printErrorMessage("Response Error");
					if (data == null) {
						System.out.println("Datawas null!");
					} else {
						System.out.println("Data was " + data[0]);
					}
					return -1;
				}
			}
		}

		// Send finish
		tmp = new short[4];
		tmp[0] = OK;
		tmp[1] = OK;
		tmp[2] = OK;
		tmp[3] = OK;
		if (!writeData(tmp)) {
			printErrorMessage("Timeout sending Finish");
			return -1;
		}
		data = readData(1);
		if ((data == null) || (data[0] != OK)) {
			printErrorMessage("Response Error");
			if (data == null) {
				System.out.println("Finish was null!");
			} else {
				System.out.println("Finish was " + data[0]);
			}
			return -1;
		}
		return 0;
	}

	/**
	 * Close the serial port again.
	 */
	public void closeSerialPort() {
		HelperUtility.closePort();
	}

	private void printErrorMessage(String s) {
		System.out.println("SerialHelper: " + s);
		frame.errorMessage("Serial Error", s);
	}

	private boolean writeData(short[] data) {
		// write data. return true if success
		long startdate = (new Date()).getTime();

		SerialWriteThread t = new SerialWriteThread(data);
		t.start();
		while (!t.dataWasSent()) {
			if ((new Date()).getTime() >= (startdate + (data.length * 1000))) {
				// More than (length * 1000) milliseconds went by
				// => 1 second per byte
				return false;
			}
		}
		return true;
	}

	private short[] readData(int length) {
		// return data read or null if timeout
		long startdate = (new Date()).getTime();

		SerialReadThread t = new SerialReadThread(length);
		t.start();
		while (!t.dataIsReady()) {
			if ((new Date()).getTime() >= (startdate + (length * 1000))) {
				// More than (length * 1000) milliseconds went by
				// => 1 second per byte
				return null;
			}
		}
		return t.getSerialData();
	}
}